drivers/sensor/mcp9600/mcp9600.c

```diff
@@ -0,0 +1,119 @@
+/*
+ * Copyright (c) 2020 SER Consulting, LLC
+ * Copyright (c) 2019 Peter Bigot Consulting, LLC
+ * Copyright (c) 2016 Intel Corporation
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <errno.h>
+#include <kernel.h>
+#include <drivers/i2c.h>
+#include <init.h>
+#include <sys/byteorder.h>
+#include <sys/__assert.h>
+#include <logging/log.h>
+#include "mcp9600.h"
+
+#define DT_DRV_COMPAT microchip_mcp9600
+
+#define MCP9600_BUS_I2C DT_ANY_INST_ON_BUS_STATUS_OKAY(i2c)
+
+LOG_MODULE_REGISTER(MCP9600, CONFIG_SENSOR_LOG_LEVEL);
+
+#if DT_NUM_INST_STATUS_OKAY(DT_DRV_COMPAT) == 0
+#warning "MCP9600 driver enabled without any devices"
+#endif
+
+static int mcp9600_reg_read(const struct device *dev, uint8_t start,
+			uint8_t *buf, int size)
+{
+	struct mcp9600_data *data = dev->data;
+	const struct mcp9600_config *cfg = dev->config;
+
+	return i2c_burst_read(data->i2c_master, cfg->i2c_addr,
+						start, buf, size);
+}
+
+static int mcp9600_sample_fetch(const struct device *dev,
+			enum sensor_channel chan)
+{
+	struct mcp9600_data *data = dev->data;
+	uint8_t buf[2];
+	int size = 2;
+	int ret;
+
+	__ASSERT_NO_MSG(chan == SENSOR_CHAN_ALL);
+
+	ret = mcp9600_reg_read(dev, MCP9600_REG_TEMP_HOT, buf, size);
+	if (ret < 0) {
+		return ret;
+	}
+	data->ttemp = (buf[0] << 8) | buf[1];
```

Comment:
If you make this:

    data->ttemp = (int16_t)((buf[0] << 8) | buf[1]);

and declare `ttemp` as signed you get rid of the cast when the value is extracted, and store the value in format closer to ready-to-use (Cel/16).  You could do `(int16_t)(uint16_t)((buf[0] << 8) | buf[1])` if necessary (shouldn't be, except again for safety-related code guidelines).

Reply:
Never trusted compilers to not sign extend when they shouldn't and to not do it when they should. I prefer any of the more explicit methods. I think there is a micro in zephyr to do the byte swap as well. Would have to go through the other drivers to find it again.

Reply:
actually made it a int32, moved the sign change and multiply up so I could use a value of 1 to invalidate the value on a read failure. There actually is a safety component to my use of this device.

Reply:
tested